src/emsarray/operations/cache.py

```diff
@@ -0,0 +1,120 @@
+"""
+Operations for making caching keys for a given dataset.
+"""
+import hashlib
+import marshal
+
+import numpy
+import xarray
+
+import emsarray
+
+
+def hash_attributes(hash: "hashlib._Hash", attributes: dict) -> None:
+    """
+    Updates the provided hash with with a marshal serialised byte representation of the given attribute dictionary.
+
+    Parameters
+    ----------
+    hash : hashlib-style hash instance
+        The hash instance to update with the given attribute dict.
+        This must follow the interface defined in :mod:`hashlib`.
+    attributes: dict
+        Expects a marshal compatible dictionary.
+    """
+    # Prepend the marshal encoding version
+    marshal_version = 4
+    hash_int(hash, marshal_version)
+    # Specify marshal encoding version when serialising
+    attribute_dict_marshal_bytes = marshal.dumps(attributes, marshal_version)
+    # Prepend the number of attributes
+    hash_int(hash, len(attributes))
+    # Prepend the size of the pickled attributes
+    hash_int(hash, len(attribute_dict_marshal_bytes))
+    hash.update(attribute_dict_marshal_bytes)
+
+
+def hash_string(hash: "hashlib._Hash", value: str) -> None:
+    """
+    Updates the provided hash with with a utf-8 encoded byte representation of the provided string.
+
+    Parameters
+    ----------
+    hash : hashlib-style hash instance
+        The hash instance to update with the given attribute dict.
+        This must follow the interface defined in :mod:`hashlib`.
+    attributes: str
+        Expects a string that can be encoded in utf-8.
+    """
+    # Prepend the str length
+    hash_int(hash, len(value))
+    hash.update(value.encode('utf-8'))
+
+
+def hash_int(hash: "hashlib._Hash", value: int) -> None:
+    """
+    Updates the provided hash with an encoded byte representation of the provided int.
+
+    Parameters
+    ----------
+    hash : hashlib-style hash instance
+        The hash instance to update with the given attribute dict.
+        This must follow the interface defined in :mod:`hashlib`.
+    attributes: int
+        Expects an int that can be represented in a numpy int32.
+    """
+    hash.update(numpy.int32(value).tobytes())
```

Comment:
One final bit of paranoia: lets protect against integer overflows here. If someone needs to hash an int bigger than an int32 we can reevaluate our methodology, but until that point we should protect against silently discarding the high bits of this number. This is only used currently to hash the length of strings and the size of coordinate arrays, so an upper bound on 32 bit integers feels reasonable for the current use cases.

```python
with numpy.errstate(over='raise'):
    hash.update(numpy.int32(value).tobytes())
```

This can be tested with:

```python
# this is the maximum 32 bit integer, no error
hash_int(hashlib.sha1(), 2 ** 31 - 1)
# this should overflow
with pytest.raises(OverflowError):
    hash_int(hashlib.sha1(), 2 ** 31)
```